File: java/kudu-client/src/main/java/org/apache/kudu/client/ResourceMetrics.java
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.apache.kudu.client;

import static com.google.protobuf.Descriptors.FieldDescriptor.JavaType;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.LongAdder;
import java.util.stream.Collectors;

import com.google.common.base.Preconditions;
import com.google.protobuf.Descriptors.FieldDescriptor;

import org.apache.kudu.tserver.Tserver.ResourceMetricsPB;
import org.apache.yetus.audience.InterfaceAudience;
import org.apache.yetus.audience.InterfaceStability;

/**
 * A container for scanner resource metrics.
 * <p>
 * This class wraps a mapping from metric name to metric value for server-side
 * metrics associated with a scanner.
 */
@InterfaceAudience.Public
@InterfaceStability.Evolving
public class ResourceMetrics {
  private Map<String, LongAdder> metrics = new ConcurrentHashMap<>();

  /**
   * Returns a copy of this ResourceMetrics's underlying map of metric name to
   * metric value.
   * @return a map of metric name to metric value
   */
  public Map<String, Long> get() {
      return metrics.entrySet().stream()
              .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().sum()));
  }

  /**
   * Returns the value of the metric named by 'name', or 0 if there is no such metric.
   * @param name the name of the metric to get the value for
   * @return the value of the named metric; if the metric is not found, returns 0
   */
  public long getMetric(String name) {
      return metrics.getOrDefault(name, new LongAdder()).sum();
  }

  /**
   * Increment this instance's metric values with those found in 'resourceMetricsPb'.
   * @param resourceMetricsPb resource metrics protobuf object to be used to update this object
   */
  void update(ResourceMetricsPB resourceMetricsPb) {
    Preconditions.checkNotNull(resourceMetricsPb);
    for (Map.Entry<FieldDescriptor, Object> entry : resourceMetricsPb.getAllFields().entrySet()) {
      FieldDescriptor field = entry.getKey();
      if (field.getJavaType() == JavaType.LONG) {
        increment(field.getName(), (Long) entry.getValue());
      }
    }
  }

  /**
   * Increment the metric value by the specific amount.
   * @param name the name of the metric whose value is to be incremented
   * @param amount the amount to increment the value by
   */
  private void increment(String name, long amount) {
    metrics.computeIfAbsent(name, k -> new LongAdder()).add(amount);
  }
}

File: src/kudu/client/resource_metrics.cc
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include "kudu/client/resource_metrics.h"

#include <cstdint>
#include <map>
#include <mutex>
#include <set>
#include <string>
#include <utility>

#include <sparsehash/dense_hash_map>

#include "kudu/client/resource_metrics-internal.h"
#include "kudu/gutil/map-util.h"
#include "kudu/gutil/strings/stringpiece.h"

namespace kudu {

class simple_spinlock;

namespace client {

ResourceMetrics::ResourceMetrics() :
  data_(new Data) {}

ResourceMetrics::~ResourceMetrics() {
  delete data_;
}

void ResourceMetrics::Increment(const std::string& name, int64_t amount) {
  data_->Increment(name, amount);
}

std::map<std::string, int64_t> ResourceMetrics::Get() const {
  return data_->Get();
}

int64_t ResourceMetrics::GetMetric(const std::string& name) const {
  return data_->GetMetric(name);
}

ResourceMetrics::Data::Data() {
  counters_.set_empty_key("");
}

ResourceMetrics::Data::~Data() {}

void ResourceMetrics::Data::Increment(const std::string& name, int64_t amount) {
  std::lock_guard<simple_spinlock> l(lock_);
  auto it = owned_strings_.insert(name).first;
  counters_[*it] += amount;
}
void ResourceMetrics::Data::Increment(StringPiece name, int64_t amount) {
  std::lock_guard<simple_spinlock> l(lock_);
  counters_[name] += amount;
}

std::map<std::string, int64_t> ResourceMetrics::Data::Get() const {
  std::map<std::string, int64_t> ret;
  std::lock_guard<simple_spinlock> l(lock_);
  for (const auto& p : counters_) {
    ret.emplace(p.first.as_string(), p.second);
  }
  return ret;
}

int64_t ResourceMetrics::Data::GetMetric(const std::string& name) const {
  std::lock_guard<simple_spinlock> l(lock_);
  return FindWithDefault(counters_, name, 0);
}

} // namespace client
} // namespace kudu
